### ISSUE TITLE
agda livecheck issue

### ISSUE THREAD

**Issue by miccal** [Homebrew/homebrew-livecheck#814]
I am having an issue in creating a `livecheckable` for `agda`. Currently:
```
-bash-5.0.17- /Users/miccal (28) [> brew livecheck agda
agda (guessed) : 2.6.1 ==> 08-2
```
For `agda`, `brew livecheck` is using `https://github.com/agda/agda.git`, which is a bit of a mess tag-wise as can be seen by running
```
git ls-remote https://github.com/agda/agda.git
```
Now, according to the [download link on the `agda` homepage](https://wiki.portal.chalmers.se/agda/Main/Download):
> The preferred way of installing Agda is from Hackage

so I thought I could use the `:hackage_strategy` in the `heuristic`. Therefore, I thought that a suitable `livecheckable` would be
```
class Agda
  livecheck do
    url "https://hackage.haskell.org/package/Agda/src/"
  end
end
```
but this results in
```
-bash-5.0.17- /Users/miccal (28) [> brew livecheck agda
Error: agda: 404 Not Found
```
Checking with the `--debug` flag results in
```
Trying with url https://hackage.haskell.org/package/Agda/src/
Error: agda: 404 Not Found
```
However,
```
-bash-5.0.17- /Users/miccal (28) [> curl -sLI https://hackage.haskell.org/package/Agda/src/
HTTP/1.1 200 OK
[SNIP]
```
but I note that
```
-bash-5.0.17- /Users/miccal (28) [> curl -sLI https://hackage.haskell.org/package/agda/src/
HTTP/1.1 404 Not Found
[SNIP]
```
so that capital `A` in the link is important.

So, in summary, I do not really understand why I am getting a `404 Not Found` error, and the only thing I can think of is that perhaps the capital `A` is being changed to a lower-case `a`?

Thank you.

**Comment by samford** [comment]
If we look at the body of the hackage strategy, we see that it expects to encounter a hyphen in the part of the URL containing the package name:

```ruby
package = ((url.split("/")[4]).split("-")[0..-2]).join("-")
```

If we look at the stable URL in the formula, the URL is like `https://hackage.haskell.org/package/Agda-2.6.1/Agda-2.6.1.tar.gz`, so the strategy expects that it will be working with `Agda-2.6.1`. With your URL it encounters `Agda`, which doesn't contain a hyphen and breaks the package-name-detection part of the strategy.

If you add a temporary print statement in the hackage strategy to see the URL it creates, it's actually `https://hackage.haskell.org/package//src/`, since the strategy wasn't able to identify the package name in your URL.

---

With this in mind there were two ways to create a working livecheckable for this formula with the currently-available features in livecheck. Adding a livecheckable with a regex to restrict matching to only the Git tags we want would technically resolve the issue.

You can use something like `regex(/^v?(\d+(?:[._]\d{1,4})+)$/i)`, which restricts the length of the numeric parts of the version to 4 digits, which is one way to avoid the releases like `v2.6.0.1.20200307`. That said, I'm not overly fond of arbitrarily restricting numeric portions of the version (though we've had to do it in one or two livecheckables above).

However, the stable archive in the formula comes from Hackage, so it's more appropriate to check that source for new versions. In this case, there's no need to manually specify a URL and we can simply use `url :stable` to make livecheck use the stable URL from the formula, which will use the Hackage strategy. I opened a PR (#815) which does this and will close this issue once merged.

---

Beyond this, I happen to be in the process of creating a PR that approaches the package-name-finding parts of the strategies in a more defensive manner (instead of making assumptions that may fail), so this may improve things a bit. I'm also in the process of creating a PR that overhauls the debug output (while also improving JSON output), so it should be much easier to understand what's going on in livecheck.

Furthermore, in the forthcoming PR for the using/strategy feature, I'll be extending the debug output to also display the URL and regex used _within_ the strategy, so livecheck's behavior will be even more transparent.

The improvements to the debug output will make it easier to understand situations like these in the future but you can also temporarily add debug statements in livecheck yourself to understand that's going on in the interim time (as above).

**Comment by miccal** [comment]
Thank you for the detailed explanation @samford, now I understand.

> the formula URLs you can reference in the livecheckable's `url` with a symbol are `:head`, `:stable`, and `:homepage`.

 Good to know, thanks.

**Comment by miccal** [comment]
@samford a question if I may -- i notice that running `brew livecheck agda --debug` yields
```
Trying with url https://hackage.haskell.org/package/Agda-2.6.1/Agda-2.6.1.tar.gz
2.6.1 => #<Version:0x00007f9eea1f7a58 @version="2.6.1", @tokens=[#<Version::NumericToken 2>, #<Version::NumericToken 6>, #<Version::NumericToken 1>]>
agda : 2.6.1 ==> 2.6.1
```
but would this now imply that `livecheck` will only show is there is an update for `agda` if the Formula for `agda` is up-to-date?

**Comment by samford** [comment]
By default livecheck outputs information for formulae where a newer version isn't available. Livecheck only omits displaying information for these formulae if the `--newer-only` flag is passed. The output is as follows:

```
$ brew livecheck agda
agda : 2.6.1 ==> 2.6.1
```

```
$ brew livecheck --newer-only agda
```

Does this answer your question or did I misunderstand?

**Comment by miccal** [comment]
Not quite, sorry -- in this case how would `livecheck` know if a newer version of `agda` is available from upstream?

Apologies for the questions, I just like to understand these things :)

**Comment by samford** [comment]
> how would `livecheck` know if a newer version of `agda` is available from upstream?

If you take a look at [the Hackage strategy in `livecheck_strategy.rb`](https://github.com/Homebrew/homebrew-livecheck/blob/master/livecheck/livecheck_strategy.rb#L128-L133) and walk through it manually, you'll see that livecheck is checking the following URL: https://hackage.haskell.org/package/Agda/src/

The upstream page isn't specific to a given version and simply reflects the newest version of the package. When a new version is released, this page will update to reflect it and livecheck will report that version when it checks.

Some sources contain information for multiple versions (e.g., index pages) and some sources only contain information for the latest version (as here with the Hackage strategy).

**Comment by miccal** [comment]
Ah, I see now. Thank you again for your thoughtful and detailed explanation.
